charmhub/selector/selector.go

```diff
@@ -0,0 +1,352 @@
+	// Order the channelMap by the ordered supported controller series. That
+	// way we'll always find the newest one first (hopefully the most
+	// supported).
+	// Then attempt to find the revision by a channel.
+	channelMap := channelMapBySeries{
+		channelMap: in,
+		series:     c.orderedSeries,
+	}
+	sort.Sort(channelMap)
+
+	return channelMap.channelMap
+}
+
+func (c *Selector) filterByArchitecture(arch string) FilterInfoChannelMapFunc {
+	return func(channelMap transport.InfoChannelMap) bool {
+		if arch == c.emptyArch {
+			return true
+		}
+
+		platformArch := channelMap.Channel.Platform.Architecture
+		return (platformArch == arch || platformArch == c.emptyArch) ||
+			isArchInPlatforms(channelMap.Revision.Platforms, arch) ||
+			isArchInPlatforms(channelMap.Revision.Platforms, ArchAll)
+	}
+}
+
+func (c *Selector) filterBySeries(series string) FilterInfoChannelMapFunc {
+	return func(channelMap transport.InfoChannelMap) bool {
+		if series == c.emptySeries {
+			return true
+		}
+
+		platformSeries := channelMap.Channel.Platform.Series
+		return (platformSeries == series || platformSeries == c.emptySeries) ||
+			isSeriesInPlatforms(channelMap.Revision.Platforms, series) ||
+			isSeriesInPlatforms(channelMap.Revision.Platforms, SeriesAll)
+	}
+}
+
+func (c *Selector) filterByArchitectureAndSeries(arch, series string) FilterInfoChannelMapFunc {
+	return func(channelMap transport.InfoChannelMap) bool {
+		return c.filterByArchitecture(arch)(channelMap) &&
+			c.filterBySeries(series)(channelMap)
+	}
+}
+
+// FilterInfoChannelMapFunc is a type alias for representing a filter function.
+type FilterInfoChannelMapFunc func(channelMap transport.InfoChannelMap) bool
+
+func filterInfoChannelMap(in []transport.InfoChannelMap, fn FilterInfoChannelMapFunc) []transport.InfoChannelMap {
+	var filtered []transport.InfoChannelMap
+	for _, channelMap := range in {
+		if !fn(channelMap) {
+			continue
+		}
+		filtered = append(filtered, channelMap)
+	}
+	return filtered
+}
+
+func locateRevisionByChannel(channelMaps []transport.InfoChannelMap, channel charm.Channel) ([]transport.InfoRevision, bool) {
+	var (
+		revisions []transport.InfoRevision
+		found     bool
+	)
+	channel = channel.Normalize()
+	for _, channelMap := range channelMaps {
+		if rev, ok := locateRevisionByChannelMap(channelMap, channel); ok {
+			revisions = append(revisions, rev)
+			found = true
+		}
+	}
+	return revisions, found
+}
+
+func listArchitectures(revisions []transport.InfoRevision) []string {
+	arches := set.NewStrings()
+	for _, rev := range revisions {
+		for _, platform := range rev.Platforms {
+			arches.Add(platform.Architecture)
+		}
+	}
+	return arches.SortedValues()
+}
+
+type channelMapIndex struct {
+	witnessed  bool
+	channelMap transport.InfoChannelMap
+}
+
+var channelRisks = []string{"stable", "candidate", "beta", "edge"}
+
+func linkClosedChannels(channelMaps []transport.InfoChannelMap) ([]transport.InfoChannelMap, error) {
```

Comment:
I wonder if we should increase cross-team collaboration to see if there are now common snap/charmhub libs that might be extracted for use across code bases.

Reply:
This is essentially only for bundles, which they don't require. They get to use the Refresh API and only that. We have to jump through hoops to get bundles to work in the same way as charms.